File: Java/src/com/ecc/gkgo1a/K12.java
package com.ecc.gkgo1a;

import java.util.Arrays;
import java.util.Comparator;
import java.util.Random;
import java.util.Scanner;
import java.util.function.Supplier;

class Person {
    String Name;
    int BirthYear;
    float Height;
    float Weight;
}

class Staff {
    String Name;
    char Gender;
    int Age;

    Staff(String name, char gender, int age) {
        Name = name;
        Gender = gender;
        Age = age;
    }
}

class Position implements Comparable<Position> {
    int X;
    int Y;
    double Distance;

    Position(int x, int y) {
        X = x;
        Y = y;
        Distance = 0.0;
    }

    Position() { }

    @Override
    public int compareTo(Position o) {
        return Double.compare(this.Distance, o.Distance);
    }
}

public class K12 extends IExecutable {
    @Override
    public void Question1() {
        Person person = new Person();
        System.out.print("名前を入力してください＞");
        Scanner reader = new Scanner(System.in);
        person.Name = reader.nextLine();
        System.out.print("生年（西暦）を入力してください＞");
        person.BirthYear = reader.nextInt();
        System.out.print("身長を入力してください＞");
        person.Height = reader.nextFloat();
        System.out.print("体重を入力してください＞");
        person.Weight = reader.nextFloat();

        System.out.println("プロファイル：");
        System.out.printf("名前：%s\n", person.Name);
        System.out.printf("生年：%d\n", person.BirthYear);
        System.out.printf("身長：%.1f\n", person.Height);
        System.out.printf("体重：%.1f\n", person.Weight);
    }

    @Override
    public void Question2() {
        Staff[] staffs = {
                new Staff("神木隆之介", 'M', 23),
                new Staff("上白石萌音", 'F', 18),
                new Staff("長澤まさみ", 'F', 28),
                };

        System.out.println("名前\t\t性別\t年齢");
        for (int i = 0; i <= 35; i++) {
            System.out.print('-');
        }
        System.out.println();

        for (var staff:
             staffs) {
            System.out.printf("%s\t%c\t%d歳\n", staff.Name, staff.Gender, staff.Age);
        }

    }

    @Override
    public void Question3() {
        Random random = new Random();
        Supplier<Integer> getRandomNumber = () -> random.nextInt(101) - 50;
        Position playerPos = new Position(getRandomNumber.get(), getRandomNumber.get());
        Position[] enemyPos = new Position[5];
        for (int i = 0; i < enemyPos.length; i++) {
            Position pos = new Position();
            pos.X = getRandomNumber.get();
            pos.Y = getRandomNumber.get();
            pos.Distance = GetDistance(playerPos, pos);
            enemyPos[i] = pos;
        }

        System.out.printf("プレーヤーの座標：x= %5.2f\t\ty= %5.2f\n\n", (float)playerPos.X, (float)playerPos.Y);
        for (int i = 0; i < enemyPos.length; i++) {
            System.out.printf("敵キャラの座標0%d：x= %5.2f\t\ty= %5.2f\t\t距離=%5.2f\n",
                    i, (float)enemyPos[i].X, (float)enemyPos[i].Y, enemyPos[i].Distance);
        }
    }

    @Override
    public void Question4() {
        Random random = new Random();
        Supplier<Integer> getRandomNumber = () -> random.nextInt(101) - 50;
        Position playerPos = new Position(getRandomNumber.get(), getRandomNumber.get());
        Position[] enemyPos = new Position[5];
        for (int i = 0; i < enemyPos.length; i++) {
            Position pos = new Position();
            pos.X = getRandomNumber.get();
            pos.Y = getRandomNumber.get();
            pos.Distance = GetDistance(playerPos, pos);
            enemyPos[i] = pos;
        }

        System.out.printf("プレーヤーの座標：x= %5.2f\t\ty= %5.2f\n\n", (float)playerPos.X, (float)playerPos.Y);
        for (int i = 0; i < enemyPos.length; i++) {
            System.out.printf("敵キャラの座標0%d：x= %5.2f\t\ty= %5.2f\t\t距離=%5.2f\n",
                    i, (float)enemyPos[i].X, (float)enemyPos[i].Y, enemyPos[i].Distance);
        }

        System.out.println("\n並び替え後");

        var sortedList = Arrays.stream(enemyPos).sorted(Comparator.naturalOrder()).toArray(Position[]::new);
        for (int i = 0; i < sortedList.length; i++) {
            System.out.printf("敵キャラの座標0%d：x= %5.2f\t\ty= %5.2f\t\t距離=%5.2f\n",
                    i, (float)sortedList[i].X, (float)sortedList[i].Y, sortedList[i].Distance);
        }
    }

    private double GetDistance(Position playerPos, Position enemyPos) {
        int diffX = playerPos.X - enemyPos.X;
        int diffY = playerPos.Y - enemyPos.Y;
        return Math.sqrt(diffX * diffX + diffY * diffY);
    }
}

File: Scala/src/main/scala/PrimaryClasses/K12.scala
package PrimaryClasses

import IExecutable.IExecutable

import scala.util.Random

class Person(name: String = "", birthYear: Int = 0, height: Float = 0.0f, weight: Float = 0.0f) {
  var Name: String = name
  var BirthYear: Int = birthYear
  var Height: Float = height
  var Weight: Float = weight
}

class Staff(name: String, gender: Char, age: Int) {
  val Name: String = name
  val Gender: Char = gender
  val Age: Int = age
}

class Position(x: Int, y: Int, distance: Double = 0.0) {
  var X: Int = x
  var Y: Int = y
  var Distance: Double = distance
}

class K12 extends IExecutable{
  override def Question1(): Unit = {
    val person = new Person
    print("名前を入力してください＞")
    person.Name = scala.io.StdIn.readLine()
    print("生年（西暦）を入力してください＞")
    person.BirthYear = scala.io.StdIn.readInt()
    print("身長を入力してください＞")
    person.Height = scala.io.StdIn.readFloat()
    print("体重を入力してください＞")
    person.Weight = scala.io.StdIn.readFloat()

    println("プロファイル")
    printf("名前：%s\n", person.Name)
    printf("生年：%d\n", person.BirthYear)
    printf("身長：%.1f\n", person.Height)
    printf("体重：%.1f\n", person.Weight)
  }

  override def Question2(): Unit = {
    val staffs = Array(
      new Staff("神木隆之介", 'M', 23),
      new Staff("上白石萌音", 'F', 18),
      new Staff("長澤まさみ", 'F', 28)
    )
    println("名前\t\t性別\t年齢")
    for (_ <- 1 to 35) {
      print('-')
    }
    println()

    for (i <- staffs.indices) {
      printf("%s\t%c\t%d\n", staffs(i).Name, staffs(i).Gender, staffs(i).Age)
    }
  }

  def GetDistance(playerPos: Position, enemyPos: Position): Double = {
    val diffX = playerPos.X - enemyPos.X
    val diffY = playerPos.Y - enemyPos.Y
    scala.math.sqrt((diffX * diffX + diffY * diffY).toDouble)
  }

  override def Question3(): Unit = {
    val getRandomNumber = () => Random.nextInt(101) - 50
    val playerPos = new Position(getRandomNumber(), getRandomNumber())
    val enemyPos = new Array[Position](5)
    for (i <- enemyPos.indices) {
      enemyPos(i) = new Position(
        getRandomNumber(),
        getRandomNumber(),
      )

      enemyPos(i).Distance = GetDistance(playerPos, enemyPos(i))
    }

    printf("プレーヤーの座標：x= %d\t\ty= %d\n", playerPos.X, playerPos.Y)
    for (i <- enemyPos.indices) {
      printf("敵キャラの座標0%d：x= %5.2f\t\ty= %5.2f\t\t距離= %5.2f\n",
        i, enemyPos(i).X.toFloat, enemyPos(i).Y.toFloat, enemyPos(i).Distance)
    }
  }

  override def Question4(): Unit = {
    val getRandomNumber = () => Random.nextInt(101) - 50
    val playerPos = new Position(getRandomNumber(), getRandomNumber())
    val enemyPos = new Array[Position](5)
    for (i <- enemyPos.indices) {
      enemyPos(i) = new Position(
        getRandomNumber(),
        getRandomNumber(),
      )

      enemyPos(i).Distance = GetDistance(playerPos, enemyPos(i))
    }

    printf("プレーヤーの座標：x= %d\t\ty= %d\n", playerPos.X, playerPos.Y)
    for (i <- enemyPos.indices) {
      printf("敵キャラの座標0%d：x= %5.2f\t\ty= %5.2f\t\t距離= %5.2f\n",
        i, enemyPos(i).X.toFloat, enemyPos(i).Y.toFloat, enemyPos(i).Distance)
    }

    println("\n並び替え後")
    val sortedPos = enemyPos.sortBy((pos: Position) => pos.Distance)
    for (i <- sortedPos.indices) {
      printf("敵キャラの座標0%d：x= %5.2f\t\ty= %5.2f\t\t距離= %5.2f\n",
        i, sortedPos(i).X.toFloat, sortedPos(i).Y.toFloat, sortedPos(i).Distance)
    }
  }
}
